File: web/src/api/modules/llm.ts
import request from '@/api/request'

export interface LLMProvider {
  id: string
  provider_key: string
  name: string
  base_url?: string | null
  status: 'active' | 'inactive'
  created_at: string
  updated_at: string
}

export interface LLMModel {
  id: string
  provider_id: string
  provider_key?: string | null
  model_key: string
  name: string
  capabilities: string[]
  context_window?: number | null
  input_price_per_1m?: number | null
  output_price_per_1m?: number | null
  status: 'active' | 'inactive'
  created_at: string
  updated_at: string
}

export interface LLMCredential {
  id: string
  provider_id: string
  provider_key?: string | null
  name: string
  secret_set: boolean
  masked_secret: string
  active: boolean
  created_at: string
  updated_at: string
}

export interface LLMRoute {
  id: string
  route_key: string
  name: string
  provider_id: string
  provider_key?: string | null
  model_id: string
  model_name?: string | null
  credential_id?: string | null
  credential_name?: string | null
  temperature: number
  max_tokens: number
  timeout_seconds: number
  fallback_route_keys: string[]
  active: boolean
  created_at: string
  updated_at: string
}

export interface LLMConfig {
  api_key_set: boolean
  default_model: string
  default_temperature: number
  max_tokens: number
  model_routes: Record<string, string>
  providers: LLMProvider[]
  models: LLMModel[]
  credentials: LLMCredential[]
  routes: LLMRoute[]
}

export const llmApi = {
  get: () => request.get<LLMConfig>('/llm'),

  update: (body: {
    default_model: string
    default_temperature: number
    max_tokens: number
    model_routes?: Record<string, string>
    api_key?: string
    vision_model?: string
    image_gen_model?: string
  }) => request.post('/llm', body),

  createProvider: (body: {
    provider_key: string
    name: string
    base_url?: string
    status?: 'active' | 'inactive'
  }) => request.post<LLMProvider>('/llm/providers', body),

  createModel: (body: {
    provider_id: string
    model_key: string
    name: string
    capabilities?: string[]
    context_window?: number
    input_price_per_1m?: number
    output_price_per_1m?: number
    status?: 'active' | 'inactive'
  }) => request.post<LLMModel>('/llm/models', body),

  createCredential: (body: {
    provider_id: string
    name: string
    secret: string
    active?: boolean
  }) => request.post<LLMCredential>('/llm/credentials', body),

  createRoute: (body: {
    route_key: string
    name: string
    provider_id: string
    model_id: string
    credential_id?: string | null
    temperature?: number
    max_tokens?: number
    timeout_seconds?: number
    fallback_route_keys?: string[]
    active?: boolean
  }) => request.post<LLMRoute>('/llm/routes', body),
}

File: src/agent_forge/models/llm.py
"""Structured LLM provider, credential, model and route models."""

from __future__ import annotations

from typing import TYPE_CHECKING

from sqlalchemy import JSON, Boolean, Float, ForeignKey, String, UniqueConstraint
from sqlalchemy.orm import Mapped, mapped_column, relationship

from .base import Base, TimestampMixin

if TYPE_CHECKING:
    pass


class LLMProviderSetting(Base, TimestampMixin):
    """A user-scoped LLM provider endpoint."""

    __tablename__ = "llm_providers"
    __table_args__ = (
        UniqueConstraint("user_id", "provider_key", name="uq_llm_providers_user_provider_key"),
    )

    id: Mapped[str] = mapped_column(String(50), primary_key=True)
    user_id: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    provider_key: Mapped[str] = mapped_column(String(60), nullable=False)
    name: Mapped[str] = mapped_column(String(120), nullable=False)
    base_url: Mapped[str | None] = mapped_column(String(500), nullable=True)
    status: Mapped[str] = mapped_column(String(30), nullable=False, default="active", index=True)
    metadata_json: Mapped[dict] = mapped_column(JSON, default=dict)

    models: Mapped[list[LLMModelSetting]] = relationship(back_populates="provider", cascade="all, delete-orphan")
    credentials: Mapped[list[LLMCredential]] = relationship(back_populates="provider", cascade="all, delete-orphan")
    routes: Mapped[list[LLMRoute]] = relationship(back_populates="provider")


class LLMModelSetting(Base, TimestampMixin):
    """A model exposed by a provider."""

    __tablename__ = "llm_models"
    __table_args__ = (
        UniqueConstraint("user_id", "model_key", name="uq_llm_models_user_model_key"),
    )

    id: Mapped[str] = mapped_column(String(50), primary_key=True)
    user_id: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    provider_id: Mapped[str] = mapped_column(ForeignKey("llm_providers.id", ondelete="CASCADE"), nullable=False)
    model_key: Mapped[str] = mapped_column(String(160), nullable=False, index=True)
    name: Mapped[str] = mapped_column(String(160), nullable=False)
    capabilities: Mapped[list[str]] = mapped_column(JSON, default=list)
    context_window: Mapped[int | None] = mapped_column(nullable=True)
    input_price_per_1m: Mapped[float | None] = mapped_column(Float, nullable=True)
    output_price_per_1m: Mapped[float | None] = mapped_column(Float, nullable=True)
    status: Mapped[str] = mapped_column(String(30), nullable=False, default="active", index=True)
    metadata_json: Mapped[dict] = mapped_column(JSON, default=dict)

    provider: Mapped[LLMProviderSetting] = relationship(back_populates="models")
    routes: Mapped[list[LLMRoute]] = relationship(back_populates="model")


class LLMCredential(Base, TimestampMixin):
    """Encrypted LLM credential. Plain secrets are never returned by the API."""

    __tablename__ = "llm_credentials"

    id: Mapped[str] = mapped_column(String(50), primary_key=True)
    user_id: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    provider_id: Mapped[str] = mapped_column(ForeignKey("llm_providers.id", ondelete="CASCADE"), nullable=False)
    name: Mapped[str] = mapped_column(String(120), nullable=False)
    encrypted_secret: Mapped[str] = mapped_column(String(4096), nullable=False)
    secret_hint: Mapped[str] = mapped_column(String(80), nullable=False)
    active: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True, index=True)
    metadata_json: Mapped[dict] = mapped_column(JSON, default=dict)

    provider: Mapped[LLMProviderSetting] = relationship(back_populates="credentials")
    routes: Mapped[list[LLMRoute]] = relationship(back_populates="credential")


class LLMRoute(Base, TimestampMixin):
    """A named model route used by stages or agents."""

    __tablename__ = "llm_routes"
    __table_args__ = (
        UniqueConstraint("user_id", "route_key", name="uq_llm_routes_user_route_key"),
    )

    id: Mapped[str] = mapped_column(String(50), primary_key=True)
    user_id: Mapped[str] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)
    route_key: Mapped[str] = mapped_column(String(80), nullable=False, index=True)
    name: Mapped[str] = mapped_column(String(120), nullable=False)
    provider_id: Mapped[str] = mapped_column(ForeignKey("llm_providers.id", ondelete="CASCADE"), nullable=False)
    model_id: Mapped[str] = mapped_column(ForeignKey("llm_models.id", ondelete="CASCADE"), nullable=False)
    credential_id: Mapped[str | None] = mapped_column(
        ForeignKey("llm_credentials.id", ondelete="SET NULL"),
        nullable=True,
    )
    temperature: Mapped[float] = mapped_column(Float, nullable=False, default=0.7)
    max_tokens: Mapped[int] = mapped_column(nullable=False, default=4096)
    timeout_seconds: Mapped[int] = mapped_column(nullable=False, default=60)
    fallback_route_keys: Mapped[list[str]] = mapped_column(JSON, default=list)
    budget_policy: Mapped[dict] = mapped_column(JSON, default=dict)
    retry_policy: Mapped[dict] = mapped_column(JSON, default=dict)
    active: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True, index=True)

    provider: Mapped[LLMProviderSetting] = relationship(back_populates="routes")
    model: Mapped[LLMModelSetting] = relationship(back_populates="routes")
    credential: Mapped[LLMCredential | None] = relationship(back_populates="routes")
